### PR TITLE
feat: add --nat flag to reth node

### LINKED ISSUE #854: Add a `--nat` flag to `reth node`
### Describe the feature

It should be possible to specify what kind of NAT mechanism to use when invoking `reth node`.

The supported methods can be found here: https://github.com/paradigmxyz/reth/blob/e6ca4c56c6a314e5553656e5b62be3bca6801b85/crates/net/nat/src/lib.rs#L25

The flag should be: `--nat <method>`, where `<method>` is any of:

- `any`
- `upnp`
- `auto` (maps to `PublicIp`)
- `extip:<ip>` (maps to `ExternalIp`)
- `none` (maps to `None`)

The default should be `any`.

Ideally, the `NatResolver` enum would be re-used for the CLI to not duplicate code.

The resolution method is set on `NetworkConfigBuilder` using:

```rs
builder::discovery(
  Discv4Config::builder()
    .external_ip_resolver(Some(nat_resolution_method))
)
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by literallymarvellous**
Fixes #854

Rough Draft

1. Reused NatResolver enum by making it pub(crate) but it threw up some errors in the file.
2. Setting this also threw up errors in other files using networkConfig builder
```
builder::discovery(
  Discv4Config::builder()
    .external_ip_resolver(Some(nat_resolution_method))
)
```

### CODE REVIEW

File: bin/reth/src/config.rs [hunk 1]
```diff
@@ -29,6 +29,7 @@ impl Config {
         chain_id: u64,
         genesis_hash: H256,
         disable_discovery: bool,
+        nat_resolution_method: NatResolver,
```
Comment (gakonst) [inline bin/reth/src/config.rs:32]: Wonder if we should move these args to the Config struct? And maybe add some clap definitions inside the Config struct and flatten them for CLI usage inside the Command struct?

File: crates/net/nat/src/lib.rs [hunk 2]
```diff
@@ -22,7 +22,7 @@ use tracing::warn;
 
 /// All builtin resolvers.
 #[derive(Debug, Clone, Copy, Eq, PartialEq, Default, Hash)]
-pub enum NatResolver {
+pub(crate) enum NatResolver {
```
Comment (mattsse) [inline crates/net/nat/src/lib.rs:25]: why making this `pub crate`?

**Review by mattsse — CHANGES_REQUESTED** [review]
question re pub crate change

**Comment by onbjerg** [comment]
Needs a rebase for the CI to pass

File: crates/net/nat/src/lib.rs (hunk 2 above)
Comment (literallymarvellous) [reply crates/net/nat/src/lib.rs:25]: I couldn't access the NatResolver from node/mod.rs, didn't wan to redefine the enum.

**Comment by literallymarvellous** [comment]
Rebase to the main branch?

**Comment by onbjerg** [comment]
Yes, we need the changes on `main` for the CI to pass

File: bin/reth/src/config.rs (hunk 1 above)
Comment (gakonst) [inline bin/reth/src/config.rs:32]:
```suggestion
        nat_resolution_method: reth_net_nat::NatResolver,
```

**Review by gakonst — APPROVED** [review]

**Comment by gakonst** [comment]
@literallymarvellous pls rebase and add missing imports. Looks good otherwise! Thank you.

File: bin/reth/src/config.rs [hunk 3]
```diff
@@ -38,6 +39,7 @@ impl Config {
             .peer_config(peer_config)
             .genesis_hash(genesis_hash)
             .chain_id(chain_id)
+            .discovery(Discv4Config::builder().external_ip_resolver(Some(nat_resolution_method)))
```
Comment (literallymarvellous) [inline bin/reth/src/config.rs:42]: I resolved the missing import but external_ip_resolver() returns &mut Discv4ConfigBuilder not Discv4ConfigBuilder. Not sure, how best to handle this.

Comment (literallymarvellous) [reply bin/reth/src/config.rs:44]: Resovled it.

**Review by gakonst — APPROVED** [review]
nice work! thanks again :D

**Pull request merged**